### ISSUE TITLE
segfault with chnset/strcpy

### ISSUE THREAD

**Issue by rorywalsh** [csound/csound#856]
I'm not quite sure what the culprit here is, but running the following .csd file causes a segfault and blocks my sound card from use until I restart my machine. I get the same issue whether running with ALSA or Jack. The code in question is from a Cabbage .csd that the author tells me used to work fine in the past (pre 6.07 I think). I've stripped back all the Cabbage stuff. 

[tgreyTestCsd.txt](https://github.com/csound/csound/files/1349209/tgreyTestCsd.txt)

**Comment by vlazzarini** [comment]
This is to do with using chnset with strings at perf time. Not sure whether it is supposed to work. But
it should not segfault.

**Comment by tgrey1** [comment]
If it helps, the code in question (stripped out of a larger project of
mine) was written and extensively tested/used in 6.08, tho I can't find a
download for that specific release and wonder if it was installed with a
previous cabbage version.  Stepping back to 6.07 has it functioning
properly again (at least on OSX), and is my short term solution for now.

Thanks for taking a look.

**Comment by vlazzarini** [comment]
I have tested the CSD here on OSX. There is no segfault, but there are two issues:

The first is:
```
new alloc for instr 1:
INIT ERROR in instr 1 (opcode bypass_gui): invalid channel name
	chnset.S	#S9	#S11	
INIT ERROR in instr 1 (opcode bypass_gui): invalid channel name
	chnset.S	#S18	#S20	
	  B  0.000 - note deleted.  i1 had 2 init errors
```
This happens because chnset first runs at i-time, and the channel name is taken from
a string at k-time (because of the k-time indexing). The clarification of access to arrays was
made (I think) at 6.08 or 6.09 and so that change might have affected this code.

I am not sure whether the semantics is right, but maybe @kunstmusik or @jpffitch can have
a look at this.

The other thing we could do is a version of chnset for strings that runs ONLY at perf-time.

2) I have a:
 WARNING: csound->Free() called with invalid pointer (0x7fd80970b580) 0 0x7fd80970b580 6d426c6b

and this is to do with the expression
```
(kbypass==1) ? "visible(0)" : "visible(1)"
```

If it happens only once in the code, it's fine. But if there are two occurrences as in the CSD,
the warning is issued. I think this might be to do with parsing. It's not chnset or strcpyk.

**Comment by vlazzarini** [comment]
I have added two opcodes: chngetks and chnsetks which will work with k-rate strings only. If you can build Csound from git, maybe you can check whether this helps. 
If it does, I will document them.

**Comment by rorywalsh** [comment]
Thanks Victor. I will try this out later and let you know.

**Comment by tgrey1** [comment]
I'd love to test it on my end too, but compiling from source will be the
first hurtle.  I'm on the road for a few more days so I'm limited on both
time and network access.  Looking forward to giving this a shot once I get
home tho.

**Comment by rorywalsh** [comment]
I'm still looking into this.

**Comment by vlazzarini** [comment]
is it working? Should I keep it and document it?

**Comment by rorywalsh** [comment]
Sorry, I didn't mean to close this. The simple tests pass, but some larger projects are still failing for me. I need to keep looking for a bit. I'll let you know.

**Comment by rorywalsh** [comment]
I'm going to need your help with this one. My simple API tests works fine, but when I try the same in Cabbage the problems persist. Consider the following example. In the first case I comment out the `chnset` line initialising the channel to a file path. Each time I change preset a file path is assigned to `gSfilePathOscA`(instr2). After this the `LoadSoundfile` instrument is called and the file is displayed. The path is printed to the console. If I then remove the commented line in instr 1, gSfilepathOscA is no longer updated?  

I also tried using `chnsetks` instead of `chnset` but that make no difference. I also tred declaring the channel first using `chn_S`. 

![out](https://user-images.githubusercontent.com/2689817/31171773-da73ed0c-a8f8-11e7-9600-6f9e0978ceaa.gif)

**Comment by vlazzarini** [comment]
is there a crash? I am asking because this does not seem to be the issue in the ticket.

**Comment by vlazzarini** [comment]
If instr 1 is running all the time, chnset will
be constantly replacing the string. That is what is happening. You haven't got a chance to change it from the API. Do you need a chnset version for i-time only? chnsetis?

**Comment by rorywalsh** [comment]
Instr 1 is only running for an single init pass. No, there is no crash,
shoe I start a new issue? I have a feeling that this issue and the one I
just closed are somehow all related.

**Comment by vlazzarini** [comment]
This does look like overwriting. Check again that instr 1 is not running when you are updating it from the API

**Comment by rorywalsh** [comment]
printing some data from instrument shows that it does indeed end after the init pass. To double check I added a call to `active` in instrument 2, it shows 0 for instrument 0 by the time instrument two starts. It has me stumped. I'll keep looking. I will start stepping through the Csound source when I get a chance and see what's happening there.

**Comment by vlazzarini** [comment]
it is strange that you said that chnsetks does not make a difference in this case. It does not run at all at i time

**Comment by rorywalsh** [comment]
I've just being stepping through things on the Csound side with the debugger and I can't find anything at fault there. Everything seems to work just as it should. The fact that my simple API tests all work add further proof that the problem must lie elsewhere, i.e., in my code. Just out of curiosity, why the `chngetks` opcode? I only ask because I've always been able to retrieve updated strings at k-rate using plain old `chnget`.

**Comment by vlazzarini** [comment]
chngetks and chnsetks are for times when you don't want them to run at i-time as in the  issue example where there was no channel
name available at i-time

Victor Lazzarini
Dean of Arts, Celtic Studies, and Philosophy
Maynooth University
Ireland

On 4 Oct 2017, at 21:05, Rory Walsh <notifications@github.com<mailto:notifications@github.com>> wrote:


I've just being stepping through things on the Csound side with the debugger and I can't find anything at fault there. Everything seems to work just as it should. The fact that my simple API tests all work add further proof that the problem must lie elsewhere, i.e., in my code. Just out of curiosity, why the chngetks opcode? I only ask because I've always been able to retrieve updated strings at k-rate using plain old chnget.

**Comment by rorywalsh** [comment]
Got it. I will try some more projects tomorrow with these and let you know
how it goes.

**Comment by tgrey1** [comment]
I had some issues I wasn't able to resolve (yet) building from the homebrew
on OSX, and I see there's talk about an upcoming release...

Might there be an OSX pre-release build I could test a little more with
before the official release?  Thanks!

**Comment by vlazzarini** [comment]
@tgrey1 Can you outline these issues?

**Comment by tgrey1** [comment]
There was nothing very specific to report.  It was a few days ago, but if I
recall the error was vague, and it directed me to a non-existant log file
for details... I tried to fiddle with it, but didn't get far, and ended up
trying to use cmake manually.  I got a CMakeOutput.log and CMakeError.log,
but I don't know how relevant this is to the brew issue.

I do want to get to the bottom of it and get brew working, but was also
hoping to get a chance to see how this issue ended up being resolved and
how it affects me before the release gets finalized.

Anyways, here's the cmake log output:

The system is: Darwin - 15.6.0 - x86_64
Compiling the C compiler identification source file "CMakeCCompilerId.c"
succeeded.
Compiler:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
Build flags:
Id flags: -c

The output was:
0


Compilation of the C compiler identification source "CMakeCCompilerId.c"
produced "CMakeCCompilerId.o"

The C compiler identification is AppleClang, found in
"/Users/tgrey/Library/Caches/Homebrew/cs6make/CMakeFiles/3.4.0/CompilerIdC/CMakeCCompilerId.o"

Compiling the CXX compiler identification source file
"CMakeCXXCompilerId.cpp" succeeded.
Compiler:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
Build flags:
Id flags: -c

The output was:
0


Compilation of the CXX compiler identification source
"CMakeCXXCompilerId.cpp" produced "CMakeCXXCompilerId.o"

The CXX compiler identification is AppleClang, found in
"/Users/tgrey/Library/Caches/Homebrew/cs6make/CMakeFiles/3.4.0/CompilerIdCXX/CMakeCXXCompilerId.o"





And here's the cmake error log:





Compiling the C compiler identification source file "CMakeCCompilerId.c"
failed.
Compiler:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
Build flags:
Id flags:

The output was:
1
Undefined symbols for architecture x86_64:
  "start", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see
invocation)


Compiling the CXX compiler identification source file
"CMakeCXXCompilerId.cpp" failed.
Compiler:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
Build flags:
Id flags:

The output was:
1
Undefined symbols for architecture x86_64:
  "start", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see
invocation)


Determining if the C compiler works failed with the following output:
Change Dir: /Users/tgrey/Library/Caches/Homebrew/cs6make/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/make" "cmTC_815d1/fast"
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f
CMakeFiles/cmTC_815d1.dir/build.make CMakeFiles/cmTC_815d1.dir/build
Building C object CMakeFiles/cmTC_815d1.dir/testCCompiler.c.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
  -isysroot
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk
 -o CMakeFiles/cmTC_815d1.dir/testCCompiler.c.o   -c
/Users/tgrey/Library/Caches/Homebrew/cs6make/CMakeFiles/CMakeTmp/testCCompiler.c
Linking C executable cmTC_815d1
/usr/local/Cellar/cmake/3.4.0/bin/cmake -E cmake_link_script
CMakeFiles/cmTC_815d1.dir/link.txt --verbose=1
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
  -isysroot
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk
-Wl,-search_paths_first -Wl,-headerpad_max_install_names
 CMakeFiles/cmTC_815d1.dir/testCCompiler.c.o  -o cmTC_815d1
Undefined symbols for architecture x86_64:
  "start", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see
invocation)
make[1]: *** [cmTC_815d1] Error 1
make: *** [cmTC_815d1/fast] Error 2

**Comment by vlazzarini** [comment]
Oh, I see. It's a unrelated issue. I thought you meant the issue here. I'll see if I can prepare a preview for you and for rory. Now I am in the middle of some work that was requested. Once that's done and settled, I'll produce a test package.

**Comment by tgrey1** [comment]
Awesome, thanks!

**Comment by jpffitch** [comment]
There seems to be more than one independent issue here.  From the messages I am unsure if the first (Rory's) still exists, and very unclear what the second one is?  Are these stoppers for a release?

**Comment by rorywalsh** [comment]
Hi John. Indications seem to appear that this is resolved. Would be best for @tgrey1 to comment and close as this issue affected his work more than anyone else.

**Comment by tgrey1** [comment]
As long as the crash/locked soundcard issue Rory had was resolved, I think
I'm happy.  The chnsetks and chngetks seem to be an acceptable workaround
for my use cases.

The change to strings and initialization seems to be a big one tho, while
rare it does introduce issues of regression for anyone (hypothetical, it's
probably no-one else but me) using strings to initialize another variable.
At the very least it seems like it needs to be covered in the
documentation, I took a quick glance and this didn't seem mentioned at all.

But I'm ok with the current state for release on anything discussed in this
thread based on my use of "6.10 (double samples) Nov 14 2017"
